feat: add case insensitive comparison for strings (ASCII only)

ammkrn opened the pull request:
Adds ASCII-only case insensitive comparison to `Char` and `String`.

fgdorais commented:
I will review this in a couple of days. Regarding names, I think the core style would prefer `ASCII` over `Ascii` (though counterexamples exist, e.g `Json` vs `JSON`). 

PS: If you need Unicode locale-independent case folding then please add a feature request to [UnicodeBasic](https://github.com/fgdorais/lean4-unicode-basic). I'm planning to add a few more normalization features this summer and this is an easy addition to the pile.

ammkrn commented:
I implemented the requested changes with one modification: the `String.Iterator` does not seem to have an instance of `Stream` (I was getting a 'failed to synthesize...' error). There's one for `Substring` which is accessible via `ToStream.toStream`, so I used that.

fgdorais commented:
Looks good!

After much consideration, the quotient type `CaseInsensitiveAsciiOnly` is not a good fit for Batteries. There are two primary reasons:

1. It has very limited functionality and adding more functions is tricky since it requires lifting.
2. It is not very efficient since most functions end up applying `caseFoldAsciiOnly` and then another string function.

In fact, I would recommend using the subtype `{ s : String // s = s.caseFoldAsciiOnly }` instead of the quotient type.

Please delete the `CaseInsensitiveAsciiOnly` file and we will be ready to merge.

fgdorais commented:
@ammkrn  Have you tried the subtype approach I suggested? I'm curious to hear whether that works for your project.

ammkrn commented:
> @ammkrn Have you tried the subtype approach I suggested? I'm curious to hear whether that works for your project.

I have not, but I have removed the requested files and squashed the commits.

fgdorais commented on Batteries/Data/String/Lemmas.lean:
```diff
@@ -728,6 +728,26 @@ theorem takeWhileAux_of_valid (p : Char → Bool) : ∀ l m r,
+  simp only [map_eq, ← data_eq_nil_iff, List.map_eq_nil_iff]
+
+@[simp]
+theorem map_isEmpty_iff (s : String) (f : Char → Char) : (s.map f).isEmpty ↔ s.isEmpty := by
```
```suggestion
theorem map_isEmpty_eq_isEmpty (s : String) (f : Char → Char) : (s.map f).isEmpty = s.isEmpty := by
```

fgdorais commented on Batteries/Data/String/Lemmas.lean:
```diff
@@ -728,6 +728,26 @@ theorem takeWhileAux_of_valid (p : Char → Bool) : ∀ l m r,
+
+@[simp]
+theorem map_isEmpty_eq_isEmpty (s : String) (f : Char → Char) : (s.map f).isEmpty = s.isEmpty := by
+  simp only [isEmpty_iff, map_eq_empty_iff]
```
```suggestion
  propext; simp only [isEmpty_iff, map_eq_empty_iff]
```

fgdorais commented:
I fixed a tiny thing: = is better than iff for `Bool` expressions.

Our Mathlib check is unreliable right now. I will have to do a manual check just to be sure. I'm juggling several pots, pans and plates right now, so please ping me again if that check takes more than a couple of weeks.

Thanks for your patience!

fgdorais commented:
Local Mathlib tests have passed. Since it's unlikely that this PR would break Mathlib, this is strong enough evidence that the alleged failure is a malfunction of our CI. I am proceeding with the merge.
